backend/cfg/cfg.ml

```diff
@@ -499,18 +506,21 @@ let is_pure_basic : basic -> bool = function
        local stack slots nor calls). *)
     false
 
+let same_location (r1 : Reg.t) (r2 : Reg.t) =
+  Reg.same_loc r1 r2
+  &&
+  match r1.loc with
+  | Unknown -> false
+  | Reg _ -> Proc.register_class r1 = Proc.register_class r2
+  | Stack _ -> Proc.stack_slot_class r1.typ = Proc.stack_slot_class r2.typ
+
 let is_noop_move instr =
   match instr.desc with
-  | Op (Move | Spill | Reload) -> (
-    match instr.arg.(0).loc with
-    | Unknown -> false
-    | Reg _ ->
-      Reg.same_loc instr.arg.(0) instr.res.(0)
-      && Proc.register_class instr.arg.(0) = Proc.register_class instr.res.(0)
-    | Stack _ ->
-      Reg.same_loc instr.arg.(0) instr.res.(0)
-      && Proc.stack_slot_class instr.arg.(0).typ
-         = Proc.stack_slot_class instr.res.(0).typ)
+  | Op
+      ( Move | Spill | Reload
+      | Scalarcast (V128_of_scalar Float64x2 | V128_to_scalar Float64x2)
+      | Vectorcast _ ) ->
```

Comment:
Yeah, vector casts are only allowed between two vectors of the same size, which always have the same register class. Floats and vectors will also always have the same class on x86, though technically that isn't guaranteed for other architectures, so maybe we should remove that part.

We did agree that `Move` will never cross classes, but the explicit cast operations can of course.

Reply:
I am easily confused: (current) vectors
and float have register class, but different
stack classes. In this particular place, that
does not matter too much because we
cannot have both operands on the
stack anyway.